Sources/Controllers/2.0/OASelectSubcategoryViewController.mm

```diff
@@ -7,214 +7,285 @@
+- (void)selectDeselectGroup:(id)sender
 {
-    OAMultiselectableHeaderView *headerView = (OAMultiselectableHeaderView *)sender;
-    NSInteger section = headerView.section;
-    NSInteger rowsCount = [self.tableView numberOfRowsInSection:section];
-    
-    [self.tableView beginUpdates];
-    if (value)
+    BOOL shouldSelect = _selectedItems.count == 0;
+    if (!shouldSelect)
+        [_selectedItems removeAllObjects];
+    else
     {
-        for (int i = 0; i < rowsCount; i++)
-            [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:section] animated:YES scrollPosition:UITableViewScrollPositionNone];
+        [_selectedItems addObjectsFromArray:_items];
+        [_selectedItems removeObjectAtIndex:0];
     }
-    else
+
+    for (NSInteger i = 1; i < _items.count; i++)
     {
-        for (int i = 0; i < rowsCount; i++)
-            [self.tableView deselectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:section] animated:YES];
+        if (shouldSelect)
+            [self.tableView selectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0] animated:NO scrollPosition:UITableViewScrollPositionNone];
+        else
+            [self.tableView deselectRowAtIndexPath:[NSIndexPath indexPathForRow:i inSection:0] animated:NO];
     }
+    [self.tableView beginUpdates];
+    [self.tableView headerViewForSection:0].textLabel.text = [[NSString stringWithFormat:OALocalizedString(@"selected_of"), (int)_selectedItems.count, _items.count - 1] upperCase];
     [self.tableView endUpdates];
+    [self.tableView reloadSections:[NSIndexSet indexSetWithIndex:0] withRowAnimation:UITableViewRowAnimationNone];
+}
+
+- (void)selectDeselectItem:(NSIndexPath *)indexPath
+{
+    if (indexPath.row > 0)
+    {
+        [self.tableView beginUpdates];
+        OAPOIUIFilter *filter = _items[indexPath.row];
+        if ([_selectedItems containsObject:filter])
+            [_selectedItems removeObject:filter];
+        else
+            [_selectedItems addObject:filter];
+        [self.tableView headerViewForSection:indexPath.section].textLabel.text = [[NSString stringWithFormat:OALocalizedString(@"selected_of"), (int) _selectedItems.count, _items.count - 1] upperCase];
+        [self.tableView endUpdates];
+        [self.tableView reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:0 inSection:indexPath.section], indexPath] withRowAnimation:UITableViewRowAnimationNone];
+    }
 }
 
 #pragma mark - UITableViewDataSource
 
--(CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section
+-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
 {
-    return [OAPOISearchHelper getHeightForFooter];
+    NSObject *item = _items[indexPath.row];
+    NSString *cellType = NSStringFromClass(item.class);
+    if ([cellType isEqualToString:kCellTypeSelectionButton])
+    {
+        static NSString * const identifierCell = kCellTypeSelectionButton;
```

Comment:
There's absolutely no need to create an extra variable here, simply use the constant. I know you copied this incorrect code, but still, we should not repeat the same mistakes.)